core/blockchain.go

```diff
@@ -1139,18 +1159,45 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
 		//   2. The block is stored as a sidechain, and is lying about it's stateroot, and passes a stateroot
 		// 	    from the canonical chain, which has not been verified.
 		// Skip all known blocks that are behind us
-		current := bc.CurrentBlock().NumberU64()
-		for block != nil && err == ErrKnownBlock && current >= block.NumberU64() {
+		var (
+			current  = bc.CurrentBlock()
+			localTd  = bc.GetTd(current.Hash(), current.NumberU64())
+			externTd = bc.GetTd(block.ParentHash(), block.NumberU64()-1) // The first block can't be nil
+		)
+		externTd = new(big.Int).Add(externTd, block.Difficulty())
+		for block != nil && err == ErrKnownBlock && localTd.Cmp(externTd) >= 0 {
 			stats.ignored++
 			block, err = it.next()
+
+			if block != nil {
+				externTd = new(big.Int).Add(externTd, block.Difficulty())
+			}
+		}
+		// The remaining blocks are still known blocks, the only scenario here is:
+		// During the fast sync, the pivot point is already submitted but rollback
+		// happens. Then node resets the head full block to a lower height via `rollback`
+		// and leaves a few known blocks in the database.
+		//
+		// When node runs a fast sync again, it can re-import a batch of known blocks via
+		// `insertChain` while a part of them have higher total difficulty than current
+		// head full block(new pivot point).
+		if err == ErrKnownBlock {
```

Comment:
I don't think you need this `err` check here. It's essentially the same as the for loop below. You can just get rid of the `if` and have the same code.